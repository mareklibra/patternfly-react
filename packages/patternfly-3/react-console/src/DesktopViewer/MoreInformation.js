import React from 'react';
import PropTypes from 'prop-types';

class MoreInformation extends React.Component {
  state = { expanded: false };

  onClick = e => {
    e.preventDefault();
    this.setState(prevState => ({ expanded: !prevState.expanded }));
  };

  render() {
    return (
      <div>
        <div className="remote-viewer-more-info">
          <a href="#" tabIndex="0" onClick={this.onClick}>
            {this.state.expanded ? <span className="fa fa-angle-down" /> : <span className="fa fa-angle-right" />}
            &nbsp; {this.props.textMoreInfo}
          </a>
        </div>
        {this.state.expanded && this.props.children}
      </div>
    );
  }
}

MoreInformation.propTypes = {
  children: PropTypes.any.isRequired,

  textMoreInfo: PropTypes.string
};

MoreInformation.defaultProps = {
  textMoreInfo: 'More Information'
};

export default MoreInformation;
