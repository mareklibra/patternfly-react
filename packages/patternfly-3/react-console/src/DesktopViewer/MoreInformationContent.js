import React from 'react';
import PropTypes from 'prop-types';

const MoreInformationInstallVariant = ({ os, content }) => (
  <li>
    <div className="more-information-install">
      <b>{os}:</b>
      {content}
    </div>
  </li>
);
MoreInformationInstallVariant.propTypes = {
  os: PropTypes.string.isRequired,
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired
};

const MoreInformationContent = () => {
  const windowsInner = (
    <div>
      Download the MSI from{' '}
      <a href="https://virt-manager.org/download/" target="_blank" rel="noopener noreferrer">
        virt-manager.org
      </a>
    </div>
  );

  return (
    <div>
      <p>
        Clicking &quot;Launch Remote Viewer&quot; will download a .vv file and launch <i>Remote Viewer</i>
      </p>
      <p>
        <i>Remote Viewer</i> is available for most operating systems. To install it, search for it in GNOME Software or
        run the following:
      </p>

      <ul>
        <MoreInformationInstallVariant os="RHEL, CentOS" content="sudo yum install virt-viewer" />
        <MoreInformationInstallVariant os="Fedora" content="sudo dnf install virt-viewer" />
        <MoreInformationInstallVariant os="Ubuntu, Debian" content="sudo apt-get install virt-viewer" />
        <MoreInformationInstallVariant os="Windows" content={windowsInner} />
      </ul>
    </div>
  );
};

export default MoreInformationContent;
