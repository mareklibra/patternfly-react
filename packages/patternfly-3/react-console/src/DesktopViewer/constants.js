// TODO: Share following constants with ConsoleSelector
export const SPICECONSOLE_TYPE = 'SpiceConsole';
export const VNCCONSOLE_TYPE = 'VncConsole';
