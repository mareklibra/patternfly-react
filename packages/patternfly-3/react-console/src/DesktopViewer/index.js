export { default as DesktopViewer } from './DesktopViewer';
