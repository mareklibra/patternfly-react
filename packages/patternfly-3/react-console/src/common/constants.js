export const NONE_TYPE = '_none_';
export const SERIAL_CONSOLE_TYPE = 'SerialConsole';
export const SPICE_CONSOLE_TYPE = 'SpiceConsole';
export const VNC_CONSOLE_TYPE = 'VncConsole';
