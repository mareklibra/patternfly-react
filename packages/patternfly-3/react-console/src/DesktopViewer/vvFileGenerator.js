import 'blob-polyfill';
import { saveAs } from 'file-saver';

import { VNC_CONSOLE_TYPE, SPICE_CONSOLE_TYPE } from '../common/constants';

export function downloadFile(fileName, content, mimeType) {
  const blob = new Blob([content], { type: mimeType });
  saveAs(blob, fileName);
}

const TYPES = {
  [VNC_CONSOLE_TYPE]: 'vnc',
  [SPICE_CONSOLE_TYPE]: 'spice'
};

export function generateVVFile({ console, type }) {
  const content =
    '[virt-viewer]\n' +
    `type=${TYPES[type] || type}\n` + // vnc or spice
    `host=${console.address}\n` +
    `port=${console.port}\n` +
    'delete-this-file=1\n' +
    'fullscreen=0\n';

  return {
    content,
    mimeType: 'application/x-virt-viewer'
  };
}
