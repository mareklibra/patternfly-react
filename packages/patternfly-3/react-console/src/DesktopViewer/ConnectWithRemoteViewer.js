import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';

import { VNCCONSOLE_TYPE, SPICECONSOLE_TYPE } from './constants';

import MoreInformation from './MoreInformation';
import MoreInformationContent from './MoreInformationContent';
import { generateVVFile, downloadFile } from './vvFileGenerator';
import consoleDetailPropType from './consoleDetailPropType';

const ConnectWithRemoteViewer = ({
  children,
  onGenerate,
  onDownload,
  spice,
  vnc,
  textConnectWithRemoteViewer,
  textMoreInfo
}) => {
  const console = spice || vnc; // strictly prefer spice over vnc
  const onClick = () => {
    const type = spice ? SPICECONSOLE_TYPE : VNCCONSOLE_TYPE;
    if (console) {
      const vv = onGenerate({ console, type });
      onDownload(vv.fileName || 'console.vv', vv.content, vv.mimeType || 'application/x-virt-viewer');
    }
  };

  return (
    <div className="remote-viewer">
      <div className="remote-viewer-launch">
        <Button bsSize="large" onClick={onClick} disabled={!console}>
          {textConnectWithRemoteViewer}
        </Button>
      </div>
      {!!console && (
        <MoreInformation textMoreInfo={textMoreInfo}>{children || <MoreInformationContent />}</MoreInformation>
      )}
    </div>
  );
};

ConnectWithRemoteViewer.propTypes = {
  children: PropTypes.any,

  spice: consoleDetailPropType,
  vnc: consoleDetailPropType,

  onGenerate: PropTypes.func,
  onDownload: PropTypes.func,

  textConnectWithRemoteViewer: PropTypes.string.isRequired,
  textMoreInfo: PropTypes.string.isRequired
};

ConnectWithRemoteViewer.defaultProps = {
  children: null /** Custom content of more-info section  */,

  spice: null,
  vnc: null,

  /** Generate content of .vv file.
   * Return an object of following format:
   * {
   *   content,  // required string value
   *   mimeType, // optional, default application/x-virt-viewer
   *   fileName  // optional, default: console.vv
   * }
   */
  onGenerate: generateVVFile,

  /** Perform file download.
   * Usually default implementation is good enough, but i.e. in case of environment with tight
   * content security policy set, this might be required.
   *
   * Examples for alternative file-download implementations:
   *   - use of iframe
   *   - use of http-server
   */
  onDownload: downloadFile
};

export default ConnectWithRemoteViewer;
