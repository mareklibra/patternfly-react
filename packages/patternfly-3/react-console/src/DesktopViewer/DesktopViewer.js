import React from 'react';
import PropTypes from 'prop-types';

import { Grid, Row, Col } from 'patternfly-react';

import ManualConnection from './ManualConnection';
import ConnectWithRemoteViewer from './ConnectWithRemoteViewer';
import consoleDetailPropType from './consoleDetailPropType';

const DesktopViewer = ({ children, spice, vnc, onGenerate, onDownload, ...props }) => (
  <Grid fluid className="desktop-viewer">
    <Row>
      <Col md={6}>
        <ConnectWithRemoteViewer spice={spice} vnc={vnc} onGenerate={onGenerate} onDownload={onDownload} {...props}>
          {children}
        </ConnectWithRemoteViewer>
      </Col>
      <Col md={6}>
        <ManualConnection spice={spice} vnc={vnc} {...props} />
      </Col>
    </Row>
  </Grid>
);

DesktopViewer.propTypes = {
  children: PropTypes.any,

  spice: consoleDetailPropType,
  vnc: consoleDetailPropType,

  onGenerate: PropTypes.func,
  onDownload: PropTypes.func,

  textManualConnection: PropTypes.string /** Internationalization */,
  textSpice: PropTypes.string,
  textVNC: PropTypes.string,
  textNoProtocol: PropTypes.string,
  textConnectWith: PropTypes.string,

  textAddress: PropTypes.string,
  textSpiceAddress: PropTypes.string,
  textVNCAddress: PropTypes.string,
  textSpicePort: PropTypes.string,
  textVNCPort: PropTypes.string,
  textSpiceTlsPort: PropTypes.string,
  textVNCTlsPort: PropTypes.string,

  textConnectWithRemoteViewer: PropTypes.string,
  textMoreInfo: PropTypes.string
};

DesktopViewer.defaultProps = {
  children: null /** Custom content of more-info section  */,

  spice: null /** Optional. Connection details for spice */,
  vnc: null /** Optional. Connection details for vnc */,

  onGenerate: undefined /** See ConnectWithRemoteViewer.defaultProps for more info */,
  onDownload: undefined /** See ConnectWithRemoteViewer.defaultProps for more info */,

  textManualConnection: 'Manual Connection',
  textSpice: 'Spice',
  textVNC: 'VNC',
  textNoProtocol: 'No connection available.',
  textConnectWith: 'Connect with any viewer application for following protocols:',

  textAddress: 'Address:',
  textSpiceAddress: 'SPICE Address:',
  textVNCAddress: 'VNC Address:',
  textSpicePort: 'SPICE Port:',
  textVNCPort: 'VNC Port:',
  textSpiceTlsPort: 'SPICE TLS Port:',
  textVNCTlsPort: 'VNC TLS Port:',

  textConnectWithRemoteViewer: 'Launch Remote Viewer',
  textMoreInfo: 'More Information'
};

export default DesktopViewer;
