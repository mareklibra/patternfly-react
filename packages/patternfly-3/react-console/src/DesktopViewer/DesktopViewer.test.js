/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { render, mount } from 'enzyme';

import DesktopViewer from './DesktopViewer';
import MoreInformationContent from './MoreInformationContent';
import { generateVVFile } from './vvFileGenerator';
import { SPICECONSOLE_TYPE } from './constants';

const spice = {
  address: 'my.host.com',
  port: 5900,
  tlsPort: '5901'
};
const vnc = {
  address: 'my.host.com',
  port: 5902,
  tlsPort: '5903'
};

test('DesktopViewer empty', () => {
  const wrapper = render(<DesktopViewer />);
  expect(wrapper).toMatchSnapshot();
});

test('DesktopViewer with Spice and VNC', () => {
  const wrapper = render(<DesktopViewer spice={spice} vnc={vnc} />);
  expect(wrapper).toMatchSnapshot();
});

test('DesktopViewer launch button', () => {
  const onDownload = jest.fn();
  const onGenerate = jest.fn().mockReturnValue({ content: 'Foo' });
  const wrapper = mount(<DesktopViewer spice={spice} vnc={vnc} onDownload={onDownload} onGenerate={onGenerate} />);
  const launchButton = wrapper.find('button');
  expect(launchButton).toHaveLength(1);
  launchButton.simulate('click');
  expect(onGenerate).toHaveBeenCalledTimes(1);
  expect(onDownload).toHaveBeenCalledTimes(1);
});

test('DesktopViewer with custom more-info content', () => {
  const wrapper = mount(
    <DesktopViewer spice={spice} vnc={vnc}>
      <p id="custom-more-info">My more-info content</p>
    </DesktopViewer>
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('#custom-more-info')).toHaveLength(0);
  const linkMoreInfo = wrapper.find('.remote-viewer-more-info > a');
  expect(linkMoreInfo).toHaveLength(1);
  linkMoreInfo.simulate('click');
  expect(wrapper.find('#custom-more-info')).toHaveLength(1);
});

test('default MoreInformationContent', () => {
  const wrapper = render(<MoreInformationContent />);
  expect(wrapper).toMatchSnapshot();
});

test('default implementation of generateVVFile()', () => {
  const output = generateVVFile({ console: spice, type: SPICECONSOLE_TYPE });
  expect(output.mimeType).toMatch('application/x-virt-viewer');
  expect(output.content).toMatch(
    '[virt-viewer]\ntype=spice\nhost=my.host.com\nport=5900\ndelete-this-file=1\nfullscreen=0\n'
  );
});
