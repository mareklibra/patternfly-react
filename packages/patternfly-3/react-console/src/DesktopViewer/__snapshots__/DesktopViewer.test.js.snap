// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DesktopViewer custom topClassName 1`] = `
<div
  class="desktop-viewer-pf my-custom-class container-fluid"
>
  <div
    class="row"
  >
    <div
      class="col-md-6"
    >
      <div
        class="remote-viewer-pf"
      >
        <div
          class="remote-viewer-pf-launch"
        >
          <button
            class="btn btn-lg btn-default"
            disabled=""
            type="button"
          >
            Launch Remote Viewer
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-md-6"
    >
      <div
        class="manual-connection-pf"
      >
        <h2>
          Manual Connection
        </h2>
        <p>
          No connection available.
        </p>
        <div>
          <dl
            class="dl-horizontal"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DesktopViewer empty 1`] = `
<div
  class="desktop-viewer-pf container-fluid"
>
  <div
    class="row"
  >
    <div
      class="col-md-6"
    >
      <div
        class="remote-viewer-pf"
      >
        <div
          class="remote-viewer-pf-launch"
        >
          <button
            class="btn btn-lg btn-default"
            disabled=""
            type="button"
          >
            Launch Remote Viewer
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-md-6"
    >
      <div
        class="manual-connection-pf"
      >
        <h2>
          Manual Connection
        </h2>
        <p>
          No connection available.
        </p>
        <div>
          <dl
            class="dl-horizontal"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DesktopViewer with Spice and VNC 1`] = `
<div
  class="desktop-viewer-pf container-fluid"
>
  <div
    class="row"
  >
    <div
      class="col-md-6"
    >
      <div
        class="remote-viewer-pf"
      >
        <div
          class="remote-viewer-pf-launch"
        >
          <button
            class="btn btn-lg btn-default"
            type="button"
          >
            Launch Remote Viewer
          </button>
        </div>
        <div
          class="expand-collapse-pf"
        >
          <div
            class="expand-collapse-pf-link-container"
          >
            <span
              class="expand-collapse-pf-separator"
            />
            <button
              class="btn btn-link"
              type="button"
            >
              <span
                aria-hidden="true"
                class="fa fa-angle-right"
              />
              More Information
            </button>
            <span
              class="expand-collapse-pf-separator"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-md-6"
    >
      <div
        class="manual-connection-pf"
      >
        <h2>
          Manual Connection
        </h2>
        <p>
          Connect with any viewer application for following protocols:
        </p>
        <div>
          <dl
            class="dl-horizontal"
          >
            <dt>
              Address:
            </dt>
            <dd>
              my.host.com
            </dd>
            <dt>
              SPICE Port:
            </dt>
            <dd>
              5900
            </dd>
            <dt>
              SPICE TLS Port:
            </dt>
            <dd>
              5901
            </dd>
            <dt>
              VNC Port:
            </dt>
            <dd>
              5902
            </dd>
            <dt>
              VNC TLS Port:
            </dt>
            <dd>
              5903
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DesktopViewer with custom more-info content 1`] = `
<DesktopViewer
  spice={
    Object {
      "address": "my.host.com",
      "port": 5900,
      "tlsPort": "5901",
    }
  }
  textAddress="Address:"
  textConnectWith="Connect with any viewer application for following protocols:"
  textConnectWithRemoteViewer="Launch Remote Viewer"
  textManualConnection="Manual Connection"
  textMoreInfo="More Information"
  textNoProtocol="No connection available."
  textSpice="Spice"
  textSpiceAddress="SPICE Address:"
  textSpicePort="SPICE Port:"
  textSpiceTlsPort="SPICE TLS Port:"
  textVNC="VNC"
  textVNCAddress="VNC Address:"
  textVNCPort="VNC Port:"
  textVNCTlsPort="VNC TLS Port:"
  topClassName=""
  vnc={
    Object {
      "address": "my.host.com",
      "port": 5902,
      "tlsPort": "5903",
    }
  }
>
  <Grid
    bsClass="container"
    className="desktop-viewer-pf"
    componentClass="div"
    fluid={true}
  >
    <div
      className="desktop-viewer-pf container-fluid"
    >
      <Row
        bsClass="row"
        componentClass="div"
      >
        <div
          className="row"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={6}
          >
            <div
              className="col-md-6"
            >
              <ConnectWithRemoteViewer
                onDownload={[Function]}
                onGenerate={[Function]}
                spice={
                  Object {
                    "address": "my.host.com",
                    "port": 5900,
                    "tlsPort": "5901",
                  }
                }
                textAddress="Address:"
                textConnectWith="Connect with any viewer application for following protocols:"
                textConnectWithRemoteViewer="Launch Remote Viewer"
                textManualConnection="Manual Connection"
                textMoreInfo="More Information"
                textNoProtocol="No connection available."
                textSpice="Spice"
                textSpiceAddress="SPICE Address:"
                textSpicePort="SPICE Port:"
                textSpiceTlsPort="SPICE TLS Port:"
                textVNC="VNC"
                textVNCAddress="VNC Address:"
                textVNCPort="VNC Port:"
                textVNCTlsPort="VNC TLS Port:"
                vnc={
                  Object {
                    "address": "my.host.com",
                    "port": 5902,
                    "tlsPort": "5903",
                  }
                }
              >
                <div
                  className="remote-viewer-pf"
                >
                  <div
                    className="remote-viewer-pf-launch"
                  >
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsSize="large"
                      bsStyle="default"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <button
                        className="btn btn-lg btn-default"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Launch Remote Viewer
                      </button>
                    </Button>
                  </div>
                  <ExpandCollapse
                    align="center"
                    bordered={false}
                    className=""
                    expanded={false}
                    textCollapsed="More Information"
                    textExpanded="More Information"
                  >
                    <div
                      className="expand-collapse-pf"
                    >
                      <div
                        className="expand-collapse-pf-link-container"
                      >
                        <span
                          className="expand-collapse-pf-separator"
                        />
                        <Button
                          active={false}
                          block={false}
                          bsClass="btn"
                          bsStyle="link"
                          disabled={false}
                          onClick={[Function]}
                        >
                          <button
                            className="btn btn-link"
                            disabled={false}
                            onClick={[Function]}
                            type="button"
                          >
                            <Icon
                              name="angle-right"
                              type="fa"
                            >
                              <FontAwesome
                                name="angle-right"
                              >
                                <span
                                  aria-hidden={true}
                                  className="fa fa-angle-right"
                                />
                              </FontAwesome>
                            </Icon>
                            More Information
                          </button>
                        </Button>
                        <span
                          className="expand-collapse-pf-separator"
                        />
                      </div>
                    </div>
                  </ExpandCollapse>
                </div>
              </ConnectWithRemoteViewer>
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={6}
          >
            <div
              className="col-md-6"
            >
              <ManualConnection
                spice={
                  Object {
                    "address": "my.host.com",
                    "port": 5900,
                    "tlsPort": "5901",
                  }
                }
                textAddress="Address:"
                textConnectWith="Connect with any viewer application for following protocols:"
                textConnectWithRemoteViewer="Launch Remote Viewer"
                textManualConnection="Manual Connection"
                textMoreInfo="More Information"
                textNoProtocol="No connection available."
                textSpice="Spice"
                textSpiceAddress="SPICE Address:"
                textSpicePort="SPICE Port:"
                textSpiceTlsPort="SPICE TLS Port:"
                textVNC="VNC"
                textVNCAddress="VNC Address:"
                textVNCPort="VNC Port:"
                textVNCTlsPort="VNC TLS Port:"
                vnc={
                  Object {
                    "address": "my.host.com",
                    "port": 5902,
                    "tlsPort": "5903",
                  }
                }
              >
                <div
                  className="manual-connection-pf"
                >
                  <h2>
                    Manual Connection
                  </h2>
                  <p>
                    Connect with any viewer application for following protocols:
                  </p>
                  <div>
                    <dl
                      className="dl-horizontal"
                    >
                      <Detail
                        title="Address:"
                        value="my.host.com"
                      >
                        <dt>
                          Address:
                        </dt>
                        <dd>
                          my.host.com
                        </dd>
                      </Detail>
                      <Detail
                        title="SPICE Port:"
                        value={5900}
                      >
                        <dt>
                          SPICE Port:
                        </dt>
                        <dd>
                          5900
                        </dd>
                      </Detail>
                      <Detail
                        title="SPICE TLS Port:"
                        value="5901"
                      >
                        <dt>
                          SPICE TLS Port:
                        </dt>
                        <dd>
                          5901
                        </dd>
                      </Detail>
                      <Detail
                        title="VNC Port:"
                        value={5902}
                      >
                        <dt>
                          VNC Port:
                        </dt>
                        <dd>
                          5902
                        </dd>
                      </Detail>
                      <Detail
                        title="VNC TLS Port:"
                        value="5903"
                      >
                        <dt>
                          VNC TLS Port:
                        </dt>
                        <dd>
                          5903
                        </dd>
                      </Detail>
                    </dl>
                  </div>
                </div>
              </ManualConnection>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Grid>
</DesktopViewer>
`;

exports[`default MoreInformationContent 1`] = `
<div>
  <p>
    Clicking "Launch Remote Viewer" will download a .vv file and launch 
    <i>
      Remote Viewer
    </i>
  </p>
  <p>
    <i>
      Remote Viewer
    </i>
     is available for most operating systems. To install it, search for it in GNOME Software or run the following:
  </p>
  <ul>
    <li>
      <div
        class="more-information-pf-install"
      >
        <b>
          RHEL, CentOS:
        </b>
        sudo yum install virt-viewer
      </div>
    </li>
    <li>
      <div
        class="more-information-pf-install"
      >
        <b>
          Fedora:
        </b>
        sudo dnf install virt-viewer
      </div>
    </li>
    <li>
      <div
        class="more-information-pf-install"
      >
        <b>
          Ubuntu, Debian:
        </b>
        sudo apt-get install virt-viewer
      </div>
    </li>
    <li>
      <div
        class="more-information-pf-install"
      >
        <b>
          Windows:
        </b>
        <div>
          Download the MSI from 
          <a
            href="https://virt-manager.org/download/"
            rel="noopener noreferrer"
            target="_blank"
          >
            virt-manager.org
          </a>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
